### PR TITLE
Remove redundant slice statements during SimplifyDefUse (#2583)

### PR DESCRIPTION
**Pull request by dbrajovic-syrmia**
* Overwritten slice statements skipped
* Improved logging in FindUninitialized

When FU is registering uses in a lhs slice of an assignment statement, it skips those
which the given slice can overwrite.

To make these changes more comprehensible, I'll start by explaining my view of the problem.

How removing redundant assignment statements works (no-slice lhs assignments):
The issue starts by pointing out that the following
```
h.data = 8;
h.data = 8;
```
gets reduced to
```
h.data = 8;
```
But these snippets are misleading - in reality, the above holds true only when
```
h.data = 8;
h.data = 8;
// a third, certain kind of statement (not necessarily right after the one above) 
```
Why is this third statement required? Because the second `h.data = 8;` would disappear just as well as the first.
This reveals the underlying notion of redundancy the compiler has: 

_Consider all no-slice lhs assignments (with no lvalue on the rhs) as redundant until proven otherwise._

The proof comes from the "third" statement finding use for the second statement (more precisely, it means reading from the same `LocationSet` as `h.data = 8;`). I use the term "floater" to describe a statement that has yet to find its use. Floaters are what is actually getting thrown out during `RemoveUnused` - statements that none other found useful. Similarly, I'll use "anchoring" to describe the moment when floaters find their use. 
 
Every assignment starts out as a floater, but only assignments whose expression(s) can read a `LocationSet`, can anchor a floater (related to the same location). This explains the _assignments with no lvalue on the rhs_ part; lvalues on the rhs read a non-empty location.

Slices are treated differently, in a sense they always read some non-empty `LocationSet` whether they're on the lhs or rhs (why this doesn't apply to no-slice expressions on the lhs is unclear to me).  Consequently, slices always anchor a floater which is why some redundant assignments get to stay.

My solution comes in the form of an additional constraint on the floaters during their time of anchoring done by another slice assignment:

- Is the floater a slice assignment?
- Will the "anchorer" (slice assignment) overwrite the floater?

If both answers are true, the floating slice assignment is not anchored for this particular use (this does **not** mean it won't find its use somewhere else).

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
The def-use analysis is relatively standard. For each statement we compute a conservative approximation of all the locations it will write to (ComputeWriteSet). Then a separate pass computes for each statement the set of locations it reads. Using these and following the program control-flow we can compute the flow of dependences: a statement is useful if it wrote a location set that some other statement reads. 

There are several things that makes this analysis harder:
- we need to do inter-procedural analysis, since a table apply call can invoke actions that can affect variables in scope.
- we need a representation of the location sets. Currently we use abstract memory locations for each base type. A struct is represented as a collection of fields. This is tricky because on procedure invocations new variables may appear.
- writes happen not only on assignments, but also on procedure calls in out arguments.

Some of the corner cases are eliminated by the execution of the SideEffects pass, which removes many constructs that would be very tricky to analyze - e.g., the order of evaluation of arguments that may have side-effects.

Our current representation of location sets does not track bits within a base type; that's why we didn't support removal of redundant slices. 

I will take a look at this PR.

File: testdata/p4_16_samples_outputs/issue2583_simplify_slice-frontend.p4 [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#include <core.p4>
+
+header Header {
+    bit<32> data;
+}
+
+parser p(packet_in pckt, out Header h) {
+    state start {
+        h.data = 32w0;
+        h.data = 32w7;
```
Comment (jphurl) [inline testdata/p4_16_samples_outputs/issue2583_simplify_slice-frontend.p4:10]: shouldn't this be simplified as well?

File: frontends/p4/simplifyDefUse.cpp [hunk 2]
```diff
@@ -553,20 +627,33 @@ class FindUninitialized : public Inspector {
...
         auto fields = storage->getField(expression->member);
+        // LOG3("storage: " << storage << "\nfields: " << fields);
         reads(expression, fields);
+        if (fields == nullptr || fields->isEmpty()) {return false;}
         registerUses(expression);
         return false;
     }
 
-    bool preorder(const IR::Slice* expression) override {
+    bool  preorder(const IR::Slice* expression) override {
```
Comment (jphurl) [inline frontends/p4/simplifyDefUse.cpp:637]: extra space

**Review by jphurl — COMMENTED** [review]
Comment on coding style: I think it is best to conform to surrounding style when adding to a file. The main style difference I see is the use of the ending `}` being different than the rest of the file. I think it is best to make the usage consistent

File: testdata/p4_16_samples_outputs/issue2583_simplify_slice2-frontend.p4 [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            default: next2;
+        }
+    }
+    state next {
+        h.data = 32w8;
+        transition final;
+    }
+    state next2 {
+        h.data[7:2] = 6w9;
+        h.data[7:2] = 6w3;
```
Comment (jphurl) [inline testdata/p4_16_samples_outputs/issue2583_simplify_slice2-frontend.p4:22]: shouldn't this be simplified as well?

File: frontends/p4/simplifyDefUse.cpp [hunk 4]
```diff
@@ -132,17 +180,19 @@ class FindUninitialized : public Inspector {
 
     Definitions* getCurrentDefinitions() const {
         auto defs = definitions->getDefinitions(currentPoint, true);
-        LOG3("FU Current point is (after) " << currentPoint << " definitions are " << defs);
+        LOG3("FU Current point is (after) " << currentPoint <<
```
Comment (hanw) [inline frontends/p4/simplifyDefUse.cpp:183]: What does the 'FU' acronym stands for? =)

Comment (jnfoster) [reply frontends/p4/simplifyDefUse.cpp:183]: Presumably "Find Uninitialized." But it could easily be confused for something else.

File: testdata/p4_16_samples_outputs/issue2583_simplify_slice-frontend.p4 (hunk 1 above)
Comment (dbrajovic-syrmia) [reply testdata/p4_16_samples_outputs/issue2583_simplify_slice-frontend.p4:10]: Yes, you're right, it should be. But this makes for a great segue into another point I wanted to make. Consider this simplification (snippet):
```
parser p(packet_in pckt, out Header h, out Header h2) {
    state start {
        h.data = 8;  // A
        h2.data = h.data; // B
        h.data = 8; // C
        h2.data = 5; // D
        transition accept;
    }
}

// turns into

parser p(packet_in pckt, out Header h, out Header h2) {
    state start {
        h.data = 32w8;
        ;
        h.data = 32w8;
        h2.data = 32w5;
        transition accept;
    }
}
```
Looks familiar. Pardon the nonsensical piece of code (hard to imagine anyone writing this), but it does expose the inner workings of the pass. Despite the lack of slices, even after the simplifying, there still seems to be a redundant statement or two. To understand why, let's remind ourselves what `RemoveUnused` actually does.

Conceptually, it's intended to skim through all the assignments in the program and remove those not found in the `HasUses::used` map. Using the terminology from before, this means "remove all the floaters who aren't anchored".
But what if the floater being removed was anchoring some earlier statement (previously a floater) ? That's exactly what happened in both examples.

C and D are saved thanks to `FindUninitialized::checkOutParameters`, so they get to stay (I won't explain how, it's not relevant to the point). B gets thrown out due to having no anchor later on (C doesn't do this, no `LocationSet` to read from), but **statement A stays because statement B was its anchor**. The `h.data` on the rhs of B makes sure of this anchoring. 

So, `RemoveUnused` does what it's told and in effect, kills off the only anchor (use) statement A was having. Thus A becomes an _orphaned_ floater - a useful statement that became redundant once we removed the (known) redundant statements.

I am not entirely sure, but there may be a workaround to this. The simplest solution is to run this pass twice, but this might not scale well (how many times is enough? maybe some recursive strategy might help). 

The core problem with this pass (I think) is that we can't detect all the redundant statements in one go. We need `RemoveUnused` to expose the orphaned floaters and then remove them or find a way to detect them before we reach the removing stage.

Comment (mbudiu-vmw) [reply testdata/p4_16_samples_outputs/issue2583_simplify_slice-frontend.p4:10]: SimplifyDefUse currently removes unused assignments, but it does not remove transitively unused assignments.
This can be fixed simply by converting the pass from a PassManager to a PassRepeated, and then it will be executed until a fixed-point is reached.

**Comment by dbrajovic-syrmia** [comment]
I suppose there are a few things that should be addressed given the latest commit. @mbudiu-vmw 's suggestion of converting the pass to `PassRepeated` does exactly what's expected given what was discussed. But in turn, as indicated by the dozen of tests whose outputs I've changed, it brings about some other possible "issues" worth mentioning.

The most obvious one is the duplicated outputs of this pass. Since the pass could run several times, warnings about (for example ) uninitialized values come up more than once. While not the most daunting difference, it certainly is a nuisance.

The other issue is the significant difference in the `-frontend`, `-midend` outputs. In these files a lot of constructs get removed because they are considered `transitively unused`. Due to my lack of understanding of what these tests were designed for (as well as the P4 compiler in general), I'm unable to decide whether these changes break the tests' functionality or not.

**Comment by mbudiu-vmw** [comment]
We have some mechanism that should prevent the same error from being emitted twice. Does it fail in this case?
Also, if some reference outputs change, the workflow is to check them manually for being correct, and then to commit the new versions together with the PR. The rule is that the emitted program must have the same semantics as the original one.

**Comment by jafingerhut** [comment]
Given the number of changed outputs, hopefully we can get some help (or have already gotten some help from, part of Travis / Github Actions test results?) from the gauntlet formal verification tool.

Given the number of assignments that were deleted, perhaps some of those test programs might achieve better test coverage if more of their outputs were used, and thus could not be correctly deleted?

Just as one example that I spot-checked, this test output https://github.com/p4lang/p4c/pull/2610/files#diff-d92c92626088f62e18df7a4f6624b324702b38fead1e57f0bcab4994fc8519c8 seems much shorter than it was before, and while it may be correct, it also seems to no longer contain some enum casts, which from the name of the test program I would guess were a motivation for creating the test program in the first place.

**Comment by mbudiu-vmw** [comment]
We should make these two changes in separate PRs.
We should have just one PR that removes recursive unused statements so we can validate that independently.

**Review by mbudiu-vmw — APPROVED** [review]
This seems sound.

File: frontends/p4/simplifyDefUse.cpp [hunk 5]
```diff
@@ -554,19 +634,31 @@ class FindUninitialized : public Inspector {
 
         auto fields = storage->getField(expression->member);
         reads(expression, fields);
+        if (fields == nullptr || fields->isEmpty()) {return false;}
```
Comment (mbudiu-vmw) [inline frontends/p4/simplifyDefUse.cpp:637]: why this change?

**Review by mbudiu-vmw — COMMENTED** [review]
I will suggest the following changes nevertheless:
- let's do the PassRepeated change in a separate PR
- As you noticed, HasUses has a useless count; please change the map to a set and delete the unused `remove` method 
- let's improve a bit the logging appearance, please replace `std::endl` in def_use.* and simplifyDefUse.* with `IndentCtl::endl`. Then change the `LOG` calls for all statements in simplifyDefUse as follows: ` LOG3("FU Visiting " << dbp(statement) << " " << statement << IndentCtl::indent);` Change `setCurrent` by adding a `LOG3(IndentCtl::unindent);`
- In the next PR, when we add PassRepeated, we have to add some support so that no warning is printed twice, probably by maintaining a set of the warnings already given.

**Comment by mbudiu-vmw** [comment]
Also, can you see whether repeating this pass makes the compilation of the large programs (switch.p4) longer? A concern was that this analysis looks expensive.

File: frontends/p4/simplifyDefUse.cpp (hunk 5 above)
Comment (dbrajovic-syrmia) [reply frontends/p4/simplifyDefUse.cpp]: Apologies in advance, it is a leftover from when I first started out examining the pass; to clear up the logs. I figured if a LocationSet was empty, no uses could be registered (at least in my test examples).

**Comment by dbrajovic-syrmia** [comment]
> Just as one example that I spot-checked, this test output https://github.com/p4lang/p4c/pull/2610/files#diff-d92c92626088f62e18df7a4f6624b324702b38fead1e57f0bcab4994fc8519c8 seems much shorter than it was before, and while it may be correct, it also seems to no longer contain some enum casts, which from the name of the test program I would guess were a motivation for creating the test program in the first place.

Yes, that specific test was the reason I wanted to address the `PassRepeated` change. Another interesting note about that test is when you compare the `-first` and `-frontend` outputs (even before this PR), the very statements the test was founded upon get removed during the simplifying (only one run).  Here's what I mean:
https://github.com/p4lang/p4c/blob/f0b37d9a26151386f3b0011f97a1eaafdd28c1ad/testdata/p4_16_samples_outputs/enumCast-first.p4#L44
https://github.com/p4lang/p4c/blob/f0b37d9a26151386f3b0011f97a1eaafdd28c1ad/testdata/p4_16_samples_outputs/enumCast-frontend.p4#L42
So, when SimplifyDefUse is converted to a PassRepeated, it picks up where it left off, which is why I was uncertain about all the other tests as well. 

> I will suggest the following changes nevertheless:
> 
> 
>     * As you noticed, HasUses has a useless count; please change the map to a set and delete the unused `remove` method
> 
>     * let's improve a bit the logging appearance, please replace `std::endl` in def_use.* and simplifyDefUse.* with `IndentCtl::endl`. Then change the `LOG` calls for all statements in simplifyDefUse as follows: ` LOG3("FU Visiting " << dbp(statement) << " " << statement << IndentCtl::indent);` Change `setCurrent` by adding a `LOG3(IndentCtl::unindent);`
> 

Fortunately, I've kept around the commit representing the original change (without the PassRepeated). I've reset this branch to that commit as well as added the requested improvements above. The logs are much more readable now.

> Also, can you see whether repeating this pass makes the compilation of the large programs (switch.p4) longer? A concern was that this analysis looks expensive.

PassManager: `bmv2/testdata/p4_14_samples/switch_20160512/switch.p4 .........................   Passed   83.93 sec`
PassRepeated: `bmv2/testdata/p4_14_samples/switch_20160512/switch.p4 .........................   Passed   89.44 sec`

Both tests were run with `make check -j16` on my virtual machine (Ubuntu 18).

Regarding the second PR, for clarification, did you mean including both the slice logic _and_ converting SimplifyDefUse to PassRepeated or just the latter?
